### PR TITLE
[yamahamusiccast]  New channel for volume in decibel

### PR DESCRIPTION
**Pull request by florian-h05**
# [yamahamusiccast] New channel for volume in decibel

# Description

This PR adds one more volume channel (personally for me, the most important one), the **volume in decibel channel**.
This channel allows you to retrieve the volume level in decibel as shown on the Yamaha device and to control the volume level in decibel (same functionality as the volumeDB channel of the Yamaha Receiver binding).
It also features some small improvements and fixes (have a look at the commit message).

Please note that some imports were changed due to the openHAB codestyle auto-format and spotless.

# Testing

Tested with RX-A750 and RX-V481D.

The built jar can be downloaded here: https://1drv.ms/u/s!Al8cKKxosMjV7gEAVWkq_VDdoCVa?e=VsvJND (outdated).

Signed-off-by: Florian Hotze <florianh_dev@icloud.com>

### CODE REVIEW

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:507]: This could be `Number:Dimensionless` with `Units.DECIBEL`. See https://www.openhab.org/docs/concepts/units-of-measurement.html.

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:1103]:
```suggestion
                    setVolumeDb(value, "zone4", host);
```

probably?

File: bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml:41]: File _src/main/resources/OH-INF/i18n/yamahamusiccast.properties_ needs to be regenerated after adding new strings. Run `mvn i18n:generate-default-translations` in the bundle folder.

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:1244]: Consider introducing a local variable for the parsed values to avoid parsing twice when out of valid range.

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:48]: It seems wildcard import might be preferable here considering the number of needed constants from this file?

File: bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml:32]: This could be a small breaking change for some users, why is this default changed?

**Review by jlaur — CHANGES_REQUESTED** [review]
Looks good! Some minor comments to consider.

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:1103]: Oops, thanks for that comment!

File: bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml
Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml:32]: I have changed this default as the README says that default is false "Sync volume across linked models (default=false)".
Furthermore, I consider false to be more practical as I don't want to automatically sync the volume by default.

File: bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml
Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml:41]: Done!

Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml:41]: @jlaur I encountered one big problem that took me quite a while to solve.
See my newest comment in this PR.

**Comment by florian-h05** [comment]
@jlaur @cweitkamp 

That might be interesting for you:

When I run `mvn i18n:generate-default-translations` maven gives the following error:
```
[ERROR] No plugin found for prefix 'i18n' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories [local (/home/florian/.m2/repository), central (https://repo1.maven.org/maven2), openhab-release (https://openhab.jfrog.io/openhab/libs-release)] -> [Help 1]
```

I solved this by modifyng the `pom.xml` in the repos root and setting the version of the `i18n-maven-plugin` to `3.2.0`. Before it was `${ohc.version}` which is currently `3.3.0-SNAPSHOT`. 

So there is a problem that the i18n plugin for 3.3.0-SNAPSHOT is not on the maven central repo??

**Comment by jlaur** [comment]
> So there is a problem that the i18n plugin for 3.3.0-SNAPSHOT is not on the maven central repo??

Now that you mention it, I believe I've experienced the same problem. I solved it by:
```
mvn org.openhab.core.tools:i18n-maven-plugin:3.2.0:generate-default-translations
```

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:507]: Sorry, I am not sure how to implement the Units.DECIBEL. Therefore I only use Number:Dimensionless.

**Comment by florian-h05** [comment]
Is that something interesting for everyone/maintainers?
Should we open an issue?

**Comment by jlaur** [comment]
@florian-h05 - unfortunately the recent merge of #12528 has caused a conflict, can you resolve this?

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: To support UoM here you could use:
```suggestion
            updateState(channel, new QuantityType<>(volumeDbState, Units.DECIBEL));
```

File: bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/resources/OH-INF/thing/thing-types.xml:103]: If you will implement UoM, this could be adapted to:
```suggestion
		<state min="-80" max="12" step="0.5" pattern="%.1f %unit%"/>
```

**Comment by florian-h05** [comment]
@jlaur 

Ready for review, I have resolved the conflict in README.

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: Float comparison with the `==` or `!=` operators is not recommended. Consider even replacing this type with something else, for example `BigDecimal` in order to use exact comparison.

Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: This comparison is only used to check whether the var had an update.
Does exact comparison matter here?

Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: The volume returned from the Yamaha device always is .0 or .5 and you are not able to set anything else.

Comment (jlaur) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: I'm not a Java expert, so I'm not very much into the details, but I would presume you can risk comparing with a value like -79.9999999997 which would not equal -80. Since your volume steps are exactly 0.5 dB, probably it would be safer to switch to `BigDecimal` where accuracy is guaranteed. See for example https://howtodoinjava.com/java-examples/correctly-compare-float-double/ or many other articles for this topic.

Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: Hmm okay, I can switch if you want me to.

I would just like to mention that the yamahareceiver binding implements volumeDB with float and does the same float comparison.

Comment (jlaur) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: There is of course a possibility that I'm wrong, so if you know that this comparison will always work on all platforms, I'm fine with it. :) Just wanted to share my concern when seeing this and checking articles online. Let me know what you prefer.

Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: > Just wanted to share my concern when seeing this and checking articles online.

So, I do not know if it works on all platforms and in every case, but float is used in the yamahareceiver binding and I did never encounter any issue with volumeDB in that binding.

> Let me know what you prefer.

I prefer the float solution as it is easier to read (no BigDecimal Constructing etc.), but I had a look at the article and I think I will implement the Threshold based comparison (chapter 2 of the article).

But currently, I am wondering if this comparison is even needed. I think I will have a look at the code and then take a decision.

Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java:624]: So I have tested much, it does not matter whether that comparison is 100% exact.

Comment (florian-h05) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java]: @jlaur 
I would leave it as it is now.

File: bundles/org.openhab.binding.yamahamusiccast/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.yamahamusiccast/README.md:53]: This should now be updated to `Number:Dimensionless`.

File: bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastHandler.java]: Sure, let's leave it. Just a small update of the README.md needed before we are ready to merge.

**Comment by florian-h05** [comment]
@jlaur 

This should be ready to merge.

**Review by jlaur — APPROVED** [review]
Thanks for this addition!

**Pull request merged**
